content/blog/2023/web-platform-tests-polyfills.md

```diff
@@ -0,0 +1,177 @@
+author: ed
+date: 2023-04-20
+tags:
+  - Article
+  - Build Tools
+  - DevOps
+  - GitHub
+  - JavaScript
+  - Testing
+image:
+  src: blog/2023/graphs-and-charts.png
+summary: |
+  Learn how to leverage Web Platform Tests to give your polyfills the confidence
+  that they are implementing upcoming browser features correctly, including a
+  comprehensive report of failing/passing tests on each change.
+---
+
+## Web Platform Tests (WPT): What Are They?
+
+The [Web Platform Tests](https://web-platform-tests.org/) (WPT) are a
+cross-browser test suite for the web platform stack, including WHATWG, W3C, and
+other specifications. They help browser vendors gauge their compatibility with
+the web platform by writing tests that can be run in all browsers. This gives
+confidence to both browser projects and web developers that they are shipping
+software that is compatible with other implementations, and that they can rely
+on the web platform to work across browsers and devices without needing extra
+layers of abstraction. [The WPT
+repository](https://github.com/web-platform-tests/wpt) also provides
+documentation, tools, and resources for test writing and reviewing.
+
+You can run all tests in your current browser by visiting
+[wpt.live](https://wpt.live) or by cloning the repository and running the `./wpt
+serve` command. The tests are organized in folders, and visiting any of the HTML
+files will run all subtests and display the results.
+
+To get a bird's eye view of the test results for upcoming versions of major
+browsers, visit [wpt.fyi](https://wpt.fyi).
+
+## WPT and Polyfills
+
+The primary users of WPT are browser vendors who want to evaluate development
+versions of their browser against the test suite. However, polyfill authors can
+also leverage WPT to run the test suite on stable browsers with their polyfill
+loaded. This is useful to measure the impact of the polyfill in a quantitative
+way that doesn't rely on manual testing.
+
+WPT includes built-in support to run with an arbitrary JavaScript file inserted
+as part of the server response. The process is explained in the [official
+documentation](https://web-platform-tests.org/running-tests/testing-polyfills.html),
+but it boils down to these steps:
+
+1. Package your polyfill as a single-file, immediately-invoked script
+2. Start WPT with `./wpt serve --inject-script=/path/to/polyfill.js`
+3. Visit any WPT and ensure the script is inlined as the first element of the
+   `<head>`. This should be obvious when inspecting the page source.
+
+At this point you should hopefully notice test scores have improved with the
+help of the injected polyfill. If something goes wrong, the WPT should also
+report tests that are not completing or are failing for some other reason.
+
+## Monitoring WPT "Coverage"
+
+An important part of polyfill development is ensuring the polyfill covers all or
+most of the WPT for a given feature to maintain compliance with the spec.
```

Comment:
This might be a place to swap `WPT` with `tests` since we're referring to the tests here and not the wider WPT project